### DISCUSSION TITLE
BROKEN for abstract unix sockets - uv_pipe_connect in unix/pipe.c does not follow abstract address format rules

### DISCUSSION THREAD

**Discussion by GrahamBarnett** [libuv/libuv#3911] [General]
void uv_pipe_connect(uv_connect_t* req,
                    uv_pipe_t* handle,
                    const char* name,
                    uv_connect_cb cb)

name can be "abc\0" or for abstract sockets "\0abc\0"

the following uv__strscpy(saddr.sun_path, name, sizeof(saddr.sun_path)); is used incorrectly ... it should memcpy and force a zero termination char to handle the abstract unix socket case

       *  abstract: an abstract socket address is distinguished (from a
          pathname socket) by the fact that sun_path[0] is a null byte
          ('\0').  The socket's address in this namespace is given by
          the additional bytes in sun_path that are covered by the
          specified length of the address structure.  (Null bytes in the
          name have no special significance.)  The name has no
          connection with filesystem pathnames.  When the address of an
          abstract socket is returned, the returned addrlen is greater
          than sizeof(sa_family_t) (i.e., greater than 2), and the name
          of the socket is contained in the first (addrlen -
          sizeof(sa_family_t)) bytes of sun_path.

**Comment by GrahamBarnett** [comment]
For reference, see python getsockaddrarg in socketmodule.c

`        {
            /* regular NULL-terminated string */
            if ((size_t)path.len >= sizeof addr->sun_path) {
                PyErr_SetString(PyExc_OSError,
                                "AF_UNIX path too long");
                goto unix_out;
            }
            addr->sun_path[path.len] = 0;

            /* including the tailing NUL */
            *len_ret = path.len + offsetof(struct sockaddr_un, sun_path) + 1;
        }
        addr->sun_family = s->sock_family;
        memcpy(addr->sun_path, path.buf, path.len);
`

**Comment by bnoordhuis** [comment]
Libuv doesn't support abstract Linux sockets out of the box (because not cross-platform) but you can create and bind the socket manually, then pass it to `uv_pipe_open()`.

You can use the `pipe_getsockname_abstract` test in test/test-pipe-getsockname.c as a how-to.

**Reply by GrahamBarnett** [reply]
Thanks .. I have passed on the information to the uvloop project, that appears from your response are using your library incorrectly

**Comment by GrahamBarnett** [comment]
uvlib have now exposed uv_pipe_connect2 that takes the address length (and therefore does not internally use strlen on the address). Do you have any objection to switching to using that, then the code will be generic?
https://github.com/libuv/libuv/commit/d843b7cf7fe77ff2c134e7de9f1ce79fbecda7a6
Thanks.

**Reply by saghul** [reply]
That's already the case, apps can use the 2 variant if they need abstract socket support.

**Reply by GrahamBarnett** [reply]
I apologise .. I was trying to link the uvloop and uvlib issues and was unclear... this is now purely a uvloop issue as they should call uv_pipe_connect2 and not uv_pipe_connect as they currently do
This is now closed wrt uvlib as uvlib has exposed uv_pipe_connect2.
Many thanks for respondng.
